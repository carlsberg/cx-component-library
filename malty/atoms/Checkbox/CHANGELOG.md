# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.2.0 (2024-05-31)

### Bug Fixes

- **checkbox:** [DSM-397] Refactor checkbox component ([#551](https://github.com/CarlsbergGBS/cx-component-library/issues/551)) ([82eb0a6](https://github.com/CarlsbergGBS/cx-component-library/commit/82eb0a62c27c1df768ae94643365942d5f4398e1))
- Remove typography provider from every component ([#383](https://github.com/CarlsbergGBS/cx-component-library/issues/383)) ([5d05680](https://github.com/CarlsbergGBS/cx-component-library/commit/5d05680e40d3da5bb3042bd8191427a70606afc3))
- reverted changes ([#396](https://github.com/CarlsbergGBS/cx-component-library/issues/396)) ([ecfcf57](https://github.com/CarlsbergGBS/cx-component-library/commit/ecfcf573863253fea16e24764b6f452c3cc4f215))
- Run auto fixable lint based on new `eslint` configuration ([#501](https://github.com/CarlsbergGBS/cx-component-library/issues/501)) ([9ee94cc](https://github.com/CarlsbergGBS/cx-component-library/commit/9ee94cc2cd407f717e62d8857b5f20a74e7bbec4))

### Features

- add tsconfig files and update package.json files for each component ([3bfc644](https://github.com/CarlsbergGBS/cx-component-library/commit/3bfc644e1cfc9dbb7cf7a0469e25fce055b53240))
- **button, checkbox:** [DSM-695] Add tests ([#685](https://github.com/CarlsbergGBS/cx-component-library/issues/685)) ([7226143](https://github.com/CarlsbergGBS/cx-component-library/commit/72261439f6c841f7b4e9c6c42ab7f13d4cbc0b41))
- change organization name ([a0aa428](https://github.com/CarlsbergGBS/cx-component-library/commit/a0aa428f43138af5707a3ff4c0e36b7c056f02e6))
- **chip:** [DSM-700] Add tests ([#695](https://github.com/CarlsbergGBS/cx-component-library/issues/695)) ([45ddade](https://github.com/CarlsbergGBS/cx-component-library/commit/45ddadea01a0c56597032694df969fc5f2058559))
- **DSM-228:** Added rating component ([#379](https://github.com/CarlsbergGBS/cx-component-library/issues/379)) ([ab47354](https://github.com/CarlsbergGBS/cx-component-library/commit/ab47354db803fb051a1f87d66ee34966698df8bd))
- **DSM-248:** Headline HTML tags ([#184](https://github.com/CarlsbergGBS/cx-component-library/issues/184)) ([305f5e0](https://github.com/CarlsbergGBS/cx-component-library/commit/305f5e0fa92d5813f4877b338ad827a8c347d452))
- **DSM-285:** table ([#442](https://github.com/CarlsbergGBS/cx-component-library/issues/442)) ([8f81f8d](https://github.com/CarlsbergGBS/cx-component-library/commit/8f81f8d3436af86fd023d9737245a76a142ad51f))
- **DSM-362:** snapshots removed & unit test updates ([#307](https://github.com/CarlsbergGBS/cx-component-library/issues/307)) ([78423eb](https://github.com/CarlsbergGBS/cx-component-library/commit/78423eb5cfe04fe8a3a7af1f5af074ac6d35e235))
- **DSM-368:** Required prop & label component ([#384](https://github.com/CarlsbergGBS/cx-component-library/issues/384)) ([983bd78](https://github.com/CarlsbergGBS/cx-component-library/commit/983bd78c8f526a8faa2df601bf918b40219820f2))
- **DSM-422:** added missing testId ([#377](https://github.com/CarlsbergGBS/cx-component-library/issues/377)) ([ecbac7a](https://github.com/CarlsbergGBS/cx-component-library/commit/ecbac7a8735d5c99067644bff8eddf4527a9a222))
- **DSM-439:** Change font-family behavior ([#470](https://github.com/CarlsbergGBS/cx-component-library/issues/470)) ([ab6dad4](https://github.com/CarlsbergGBS/cx-component-library/commit/ab6dad414c42c4343de3de6abaa4cd3ff98cfca1))
- **general:** [DSM-953] refactor icons implementation ([#892](https://github.com/CarlsbergGBS/cx-component-library/issues/892)) ([08d259f](https://github.com/CarlsbergGBS/cx-component-library/commit/08d259f476ce97e9db8ac41d5eb4773eabca0d37))
- **global:** [DSM-784] add space props in all components ([#842](https://github.com/CarlsbergGBS/cx-component-library/issues/842)) ([9f03f59](https://github.com/CarlsbergGBS/cx-component-library/commit/9f03f59e7518a99abcea06bacbee989d1ddb0465))
- **global:** migrate from bit to lerna ([a0110c8](https://github.com/CarlsbergGBS/cx-component-library/commit/a0110c8831370dc762c193b17cc593eed381f990))
- **global:** update storybook files to be compliant with v7 ([#970](https://github.com/CarlsbergGBS/cx-component-library/issues/970)) ([c61d4b4](https://github.com/CarlsbergGBS/cx-component-library/commit/c61d4b40a0755becf942ad3f28758a159f8c54e4))
- integrate visual testing ([#925](https://github.com/CarlsbergGBS/cx-component-library/issues/925)) ([35dfc1c](https://github.com/CarlsbergGBS/cx-component-library/commit/35dfc1cce6b76d5d2661d552f9fb25477d3a57f7))
- remove empty composition files ([f8173f4](https://github.com/CarlsbergGBS/cx-component-library/commit/f8173f4a2ecbf80bb7b6ffe848c023ae31819c2d))
- Use new `eslint-config-carlsberg` with streamlined way to create components ([#559](https://github.com/CarlsbergGBS/cx-component-library/issues/559)) ([a459332](https://github.com/CarlsbergGBS/cx-component-library/commit/a45933215bc0b523220743f17d988d7f1ad5fce6))
