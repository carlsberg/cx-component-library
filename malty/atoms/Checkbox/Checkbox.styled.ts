import styled, { css } from 'styled-components';

export const StyledCheckboxContainer = styled.div`
  font-family: inherit;
  display: flex;
  flex-direction: column;
`;

export const StyledError = styled.label`
  color: ${({ theme }) => theme.colors.colours.system.fail.value};
  font-size: ${({ theme }) => theme.typography.desktop.text.small_default['font-size'].value};
  font-weight: bold;
  padding-top: ${({ theme }) => theme.sizes['3xs'].value};
`;

export const StyledCheckboxLabelText = styled.span`
  color: ${({ theme }) => theme.colors.colours.default['digital-black'].value};
  margin-left: ${({ theme }) => theme.sizes['2xs'].value};
  font-size: ${({ theme }) => theme.typography.desktop.text['medium-small_default']['font-size'].value};
  line-height: ${({ theme }) => theme.typography.desktop.text['medium-small_default']['line-height'].value};
`;

export const StyledCheckboxHiddenInput = styled.input.attrs({
  type: 'checkbox'
})`
  display: none;
`;

export const StyledCheckboxDisplayInput = styled.div<{ checked?: boolean }>`
  ${({ theme }) =>
    css`
      border: ${theme.borders['border-1px--solid']['border-width'].value}
        ${theme.borders['border-1px--solid']['border-style'].value}
        ${theme.colors.colours.default['digital-black'].value};
    `}
  height: ${({ theme }) => theme.sizes.s.value};
  width: ${({ theme }) => theme.sizes.s.value};
  transition: background-color 0.25s ease-in-out;
  border-radius: 3px;
  position: relative;

  &:before {
    content: '';
    opacity: 0;
  }
  ${({ checked }) =>
    checked &&
    checked === true &&
    css`
      background-color: ${({ theme }) => theme.colors.colours.default['digital-black'].value};
      &:before {
        height: 10px;
        width: 6px;
        position: absolute;
        top: calc(50% - 1px);
        left: 50%;
        border-right: 2px solid;
        border-bottom: 2px solid;
        border-bottom-right-radius: 2px;
        border-color: ${({ theme }) => theme.colors.colours.default.white.value};
        transition: 0.25s ease-in-out;
        transition-property: opacity, border-color;
        transform: translate(-50%, -50%) rotate(45deg);
        opacity: 1;
      }
    `}
  ${({ checked }) =>
    !checked &&
    checked === undefined &&
    css`
      &:before {
        width: 10px;
        position: absolute;
        top: 50%;
        left: 50%;
        border-bottom: 2px solid;
        border-color: ${({ theme }) => theme.colors.colours.default['digital-black'].value};
        transition: 0.25s ease-in-out;
        transition-property: opacity, border-color;
        transform: translate(-50%, -50%);
        opacity: 1;
      }
    `}
`;

export const StyledCheckboxLabel = styled.label<{
  required?: boolean;
}>`
  display: flex;
  align-items: center;
  cursor: pointer;
  user-select: none;
  ${({ required }) =>
    required &&
    css`
      &::after {
        content: ' *';
        color: ${({ theme }) => theme.colors.colours.system.fail.value};
      }
    `}
  &:hover {
    ${StyledCheckboxHiddenInput}:not(:checked) + ${StyledCheckboxDisplayInput} {
      &:before {
        height: 10px;
        width: 6px;
        position: absolute;
        top: calc(50% - 1px);
        left: 50%;
        border-right: 2px solid;
        border-bottom: 2px solid;
        border-bottom-right-radius: 2px;
        transition: 0.25s ease-in-out;
        transition-property: opacity, border-color;
        transform: translate(-50%, -50%) rotate(45deg);
        border-color: ${({ theme }) => theme.colors.colours.default['digital-black'].value};
        opacity: 0.2;
      }
    }
  }
`;
