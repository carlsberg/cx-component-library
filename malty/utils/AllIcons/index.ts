export { allIcons, allIconsStoryOptions } from './AllIconsMap';
