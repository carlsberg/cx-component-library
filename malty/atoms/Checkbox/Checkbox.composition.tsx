import React from 'react';
import { Checkbox } from './Checkbox';

export const CheckedCheckbox = () => <Checkbox value="Checked checkbox" checked onValueChange={() => null} />;

export const UnckeckedCheckbox = () => (
  <Checkbox value="Unchecked checkbox" checked={false} onValueChange={() => null} />
);
