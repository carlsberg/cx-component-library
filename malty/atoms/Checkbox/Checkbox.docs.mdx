---
labels: ['react', 'typescript', 'checkbox']
description: 'A Checkbox component.'
---

import { Checkbox } from './Checkbox';

Checkbox example:

A simple `checked={true}` checkbox component

```js live
<Checkbox value="Test" checked={true} onValueChange={() => null} />
```

A simple `checked={false}` checkbox component

```js live
<Checkbox value="Test" checked={false} onValueChange={() => null} />
```
