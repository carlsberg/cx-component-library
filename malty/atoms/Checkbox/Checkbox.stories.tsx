/* eslint-disable react/destructuring-assignment */
import { Story } from '@storybook/react';
import React, { useState } from 'react';
import { Checkbox as CheckboxComponent } from './Checkbox';
import { CheckboxProps } from './Checkbox.types';

export default {
  title: 'Forms/Checkbox',
  component: CheckboxComponent,
  parameters: {
    importObject: 'Checkbox',
    importPath: '@carlsberggroup/malty.atoms.checkbox',
    variants: ['undetermined', 'checked', 'unchecked'],
  },
  argTypes: {
    value: {
      control: {
        type: 'text',
      },
      description: 'This is the value to be passed',
    },
    labelText: {
      control: {
        type: 'text',
      },
      description: 'This is the label for the checkbox',
    },
    required: {
      control: 'boolean',
      description: 'Makes the checkbox required to fill',
    },
    isIndeterminate: {
      control: 'boolean',
      description: 'Makes the checkbox Indeterminate',
    },
    error: {
      control: {
        type: 'text',
      },
      description: 'Error message below',
    },
    checked: {
      control: 'boolean',
      description: 'Checked `true` or `false`.',
    },
    onValueChange: {
      description: 'Function to be executed when checkbox state changes',
    },
  },
};

const Template: Story<CheckboxProps> = (args) => {
  const [stateChecked, setStateChecked] = useState(args.checked);
  return (
    <CheckboxComponent
      {...args}
      onValueChange={() => setStateChecked(!stateChecked)}
      checked={stateChecked}
    />
  );
};

export const Checkbox = Template.bind({});

const params = new URLSearchParams(window.location.search);
const variant = params.get('variant');

switch (variant) {
  case 'undetermined':
    Checkbox.args = {
      value: 'Undetermined',
      labelText: 'Undetermined label',
      error: '',
      required: false,
      isIndeterminate: true,
    };
    break;

  case 'unchecked':
    Checkbox.args = {
      value: 'Unchecked',
      labelText: 'Unchecked label',
      error: '',
      checked: false,
      required: false,
    };
    break;

  default:
    Checkbox.args = {
      value: 'Checked',
      labelText: 'Checked label',
      error: '',
      checked: true,
      required: false,
    };
    break;
}
