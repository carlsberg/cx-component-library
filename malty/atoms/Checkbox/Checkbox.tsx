import { globalTheme as defaultTheme } from '@carlsberggroup/malty.theme.malty-theme-provider';
import React, { useContext } from 'react';
import { ThemeContext } from 'styled-components';
import {
  StyledCheckboxContainer,
  StyledCheckboxDisplayInput,
  StyledCheckboxHiddenInput,
  StyledCheckboxLabel,
  StyledCheckboxLabelText,
  StyledError
} from './Checkbox.styled';
import { CheckboxProps } from './Checkbox.types';

export const Checkbox = ({
  id,
  value = '',
  checked = false,
  labelText,
  error,
  onValueChange,
  required = false,
  dataTestId,
  isIndeterminate,
  ...props
}: CheckboxProps) => {
  const theme = useContext(ThemeContext) || defaultTheme;

  return (
    <StyledCheckboxContainer theme={theme}>
      <StyledCheckboxLabel required={required} htmlFor={id} theme={theme}>
        <StyledCheckboxHiddenInput
          type="checkbox"
          data-testid={dataTestId}
          id={id}
          value={value}
          onChange={(e) => onValueChange(e)}
          checked={checked}
          theme={theme}
          required={required}
          // eslint-disable-next-line react/jsx-props-no-spreading
          {...props}
        />
        <StyledCheckboxDisplayInput indeterminate={isIndeterminate} checked={checked} theme={theme} />
        <StyledCheckboxLabelText data-testid={`${dataTestId}-label`} theme={theme}>
          {labelText}
        </StyledCheckboxLabelText>
      </StyledCheckboxLabel>
      {error && <StyledError theme={theme}>{error}</StyledError>}
    </StyledCheckboxContainer>
  );
};
