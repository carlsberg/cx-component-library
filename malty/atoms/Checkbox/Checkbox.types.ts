import { SpaceProps } from '@carlsberg/malty.utils.space';

export interface CheckboxProps extends React.HTMLAttributes<HTMLInputElement>, SpaceProps {
  value?: string | number;
  checked?: boolean;
  labelText?: string;
  error?: string;
  onValueChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  dataTestId?: string;
  required?: boolean;
  isIndeterminate?: boolean;
  readOnly?: boolean;
  disabled?: boolean;
  fullWidth?: boolean;
}

export type StyledCheckboxContainerProps = {
  fullWidth?: boolean;
} & SpaceProps;
