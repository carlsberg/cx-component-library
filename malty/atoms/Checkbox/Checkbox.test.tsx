import { render } from '@carlsberggroup/malty.utils.test';
import { fireEvent, screen } from '@testing-library/react';
import React from 'react';
import { Checkbox } from './Checkbox';

describe('checkbox', () => {
  it('renders elements', () => {
    const mockFn = jest.fn();
    render(<Checkbox labelText="Label text" error="Error text" value="Test value" onValueChange={mockFn} checked />);
    expect(screen.getByLabelText('Label text')).toBeInTheDocument();
    expect(screen.getByText('Error text')).toBeInTheDocument();
    expect(screen.getByDisplayValue('Test value')).toBeInTheDocument();
  });

  it('calls function on click', () => {
    const mockFn = jest.fn();
    render(<Checkbox labelText="Label text" error="Error text" value="Test value" onValueChange={mockFn} checked />);
    const checkbox = screen.getByDisplayValue('Test value');
    fireEvent.click(checkbox);
    expect(mockFn).toHaveBeenCalledTimes(1);
  });
});
