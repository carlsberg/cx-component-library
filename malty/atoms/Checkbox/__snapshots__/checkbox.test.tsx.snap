// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkbox matches snapshot 1`] = `
<span
  className="sc-bdfBQB dGZTMR"
>
  <div
    className="sc-gsTEea iCcuFk"
  >
    <label
      className="sc-gKseQn nYnaa"
    >
      <input
        checked={true}
        className="sc-eCstlR lfRepL"
        onChange={[Function]}
        type="checkbox"
        value="test"
      />
      <div
        checked={true}
        className="sc-jSgvzq goUOvp"
      />
      <span
        className="sc-hKgJUU lhWjMd"
      >
        Label text
      </span>
    </label>
  </div>
</span>
`;
